### PR TITLE
feat: Solving linear recurrence using Matrix Exponentiation (with examples).

### PR DESCRIPTION
**Pull request by AshishYUO**
#### Description of Change

feat: Solving linear recurrence using Matrix Exponentiation (with examples).
[Reference for the topic](https://www.hackerearth.com/practice/notes/matrix-exponentiation-1/).
#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes:

### CODE REVIEW

**Review by ayaankhan98 — COMMENTED** [review]
Great work @AshishYUO, Thanks for your contribution.
Just few minor suggestions
```
test.cpp: In instantiation of ‘T math::linear_recurrence_matrix::get_nth_term_of_recurrence_series(const std::vector<std::vector<T> >&, const std::vector<std::vector<T> >&, int, bool) [with T = int]’:
test.cpp:206:5:   required from here
test.cpp:159:18: error: comparison of integer expressions of different signedness: ‘int’ and ‘std::vector<int>::size_type’ {aka ‘long unsigned int’} [-Werror=sign-compare]
  159 |     if (nth_term < _base_cases.back().size() - constant_or_sum_included) {
      |         ~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
cc1plus: all warnings being treated as errors
test.cpp: In instantiation of ‘T math::linear_recurrence_matrix::get_nth_term_of_recurrence_series(const std::vector<std::vector<T> >&, const std::vector<std::vector<T> >&, int, bool) [with T = int]’:
test.cpp:206:5:   required from here
test.cpp:159:18: error: comparison of integer expressions of different signedness: ‘int’ and ‘std::vector<int>::size_type’ {aka ‘long unsigned int’} [-Werror=sign-compare]
  159 |     if (nth_term < _base_cases.back().size() - constant_or_sum_included) {
      |         ~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
cc1plus: all warnings being treated as errors
```

**Review by ayaankhan98 — COMMENTED** [review]
```
test.cpp:62:46: runtime error: signed integer overflow: 1346269 * 1346269 cannot be represented in type 'int'
test.cpp:62:30: runtime error: signed integer overflow: -1781546646 + -872784047 cannot be represented in type 'int'
```

**Review by ayaankhan98 — APPROVED** [review]
LGTM!

File: math/linear_recurrence_matrix.cpp [hunk 1]
```diff
@@ -0,0 +1,368 @@
...
+         */
+        std::vector<std::vector<T>> _res =
+            matrix_multiplication(_base_cases, _res_matrix);
+
+        return _res.back().back();
+    }
+}
+
+}  // namespace linear_recurrence_matrix
+
```
Comment (Panquesito7) [inline math/linear_recurrence_matrix.cpp:181]:
```suggestion
```

File: math/linear_recurrence_matrix.cpp [hunk 2]
```diff
@@ -0,0 +1,368 @@
...
+ * Given a recurrence relation; evaluate the value of nth term.
+ * For e.g., For fibonacci series, recurrence series is `f(n) = f(n-1) + f(n-2)`
+ * where `f(0) = 0` and `f(1) = 1`.
+ * Note that the method used only demonstrates
+ * recurrence relation with one variable (n), unlike `nCr` problem, since it has
+ * two (n, r)
+ *
+ * ### Algorithm
+ * This problem can be solved using matrix exponentiation method. For simple
+ * number exponentiation, [see algorithm
```
Comment (Panquesito7) [inline math/linear_recurrence_matrix.cpp:13]: You can use `@see` here (see for example [this algorithm](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/math/gcd_of_n_numbers.cpp#L6)).

File: math/linear_recurrence_matrix.cpp [hunk 3]
```diff
@@ -0,0 +1,368 @@
+/**
+ * @brief Evaluate recurrence relation using matrix exponentiation.
```
Comment (Panquesito7) [inline math/linear_recurrence_matrix.cpp:2]: Provide a Wikipedia link in Markdown format (if available/possible).
If there's no Wikipedia link, add another web source for algorithm explanation. 🙂

File: math/linear_recurrence_matrix.cpp [hunk 4]
```diff
@@ -0,0 +1,368 @@
...
+                _mat_c[i][j] %= mod;
+            }
+        }
+    }
+    return _mat_c;
+}
+/**
+ * @brief Returns whether matrix `mat` is a [zero
+ * matrix.](https://en.wikipedia.org/wiki/Zero_matrix)
+ * @param T template type for integer as well as floating values, default is
```
Comment (Panquesito7) [inline math/linear_recurrence_matrix.cpp:71]:
```suggestion
 * @tparam T template type for integer as well as floating values, default is
```

File: math/linear_recurrence_matrix.cpp [hunk 5]
```diff
@@ -0,0 +1,368 @@
...
+        }
+    }
+    return true;
+}
+
+/**
+ * @brief Implementation of Matrix exponentiation
+ * @details returns the matrix exponentiation `(B^n)` in `k^3 * O(log2(power))`
+ * time, where `k` is the size of matrix (k by k).
+ * @param T template type for integer as well as floating values, default is
```
Comment (Panquesito7) [inline math/linear_recurrence_matrix.cpp:92]:
```suggestion
 * @tparam T template type for integer as well as floating values, default is
```

File: math/linear_recurrence_matrix.cpp [hunk 6]
```diff
@@ -0,0 +1,368 @@
...
+        }
+        power >>= 1;
+        _mat = matrix_multiplication(_mat, _mat, mod);
+    }
+
+    return _mat_answer;
+}
+
+/**
+ * @brief Implementation of nth recurrence series.
```
Comment (Panquesito7) [inline math/linear_recurrence_matrix.cpp:130]: Missing documentation for the `T` template parameter.

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
👍 Well done!

**Review by ayaankhan98 — APPROVED** [review]

File: math/linear_recurrence_matrix.cpp [hunk 7]
```diff
@@ -0,0 +1,370 @@
...
+ * @tparam T template type for integer as well as floating values, default is
+ * long long int
+ * @param _mat_a first matrix of size n * m
+ * @param _mat_b second matrix of size m * k
+ * @returns `_mat_c` resultant matrix of size n * k
+ * Complexity: `O(n*m*k)`
+ * @note The complexity in this case will be O(n^3) due to the nature of the
+ * problem. We'll be multiplying the matrix with itself most of the time.
+ */
+template <typename T = int64_t>
```
Comment (Panquesito7) [inline math/linear_recurrence_matrix.cpp:48]: Wouldn't `template <size_t T>` be better here (same in other parts of the code)?

File: math/linear_recurrence_matrix.cpp [hunk 8]
```diff
@@ -0,0 +1,370 @@
...
+ * @namespace math
+ * @brief Mathematical algorithms
+ */
+namespace math {
+/**
+ * @namespace linear_recurrence_matrix
+ * @brief Implementation of solution for linear recurrence using matrix
+ */
+namespace linear_recurrence_matrix {
+
```
Comment (Panquesito7) [inline math/linear_recurrence_matrix.cpp:34]:
```suggestion
```

File: math/linear_recurrence_matrix.cpp [hunk 9]
```diff
@@ -0,0 +1,370 @@
...
+#include <vector>    /// for std::vector STL
+
+/**
+ * @namespace math
+ * @brief Mathematical algorithms
+ */
+namespace math {
+/**
+ * @namespace linear_recurrence_matrix
+ * @brief Implementation of solution for linear recurrence using matrix
```
Comment (Panquesito7) [inline math/linear_recurrence_matrix.cpp:31]:
```suggestion
 * @brief Functions for [Linear Recurrence Matrix](https://www.hackerearth.com/practice/notes/matrix-exponentiation-1/).
```

File: math/linear_recurrence_matrix.cpp [hunk 10]
```diff
@@ -0,0 +1,370 @@
...
+         * After matrix exponentiation, multiply with the base case to evaluate
+         * the answer. The answer is always at the end of the array.
+         */
+        std::vector<std::vector<T>> _res =
+            matrix_multiplication(_base_cases, _res_matrix);
+
+        return _res.back().back();
+    }
+}
+
```
Comment (Panquesito7) [inline math/linear_recurrence_matrix.cpp:182]:
```suggestion
```

File: math/linear_recurrence_matrix.cpp (hunk 7 above)
Comment (AshishYUO) [reply math/linear_recurrence_matrix.cpp:48]: The template is used when a user wants a preferred type (e.g., integer or floating values) when evaluating recurrence relation. It won't be anything related to the size of matrix. :)

Comment (AshishYUO) [reply math/linear_recurrence_matrix.cpp:48]: I'll update the other changes though.

Comment (Panquesito7) [reply math/linear_recurrence_matrix.cpp:48]: > when a user wants a preferred type (e.g., integer or floating values)

Ah, OK, thanks for the explanation. 🙂

File: math/linear_recurrence_matrix.cpp [hunk 11]
```diff
@@ -0,0 +1,371 @@
...
+/**
+ * @namespace math
+ * @brief Mathematical algorithms
+ */
+namespace math {
+/**
+ * @namespace linear_recurrence_matrix
+ * @brief Functions for [Linear Recurrence
+ * Matrix](https://www.hackerearth.com/practice/notes/matrix-exponentiation-1/).
```
Comment (Panquesito7) [inline math/linear_recurrence_matrix.cpp:32]:
```suggestion
 * Matrix](https://www.hackerearth.com/practice/notes/matrix-exponentiation-1/) implementation.
```

File: math/linear_recurrence_matrix.cpp [hunk 12]
```diff
@@ -0,0 +1,371 @@
...
+    std::vector<std::vector<int64_t>> pell_recurrence = {{0, 1}, {1, 2}},
+                                      pell_base_case = {
+                                          {2, 2}};  // `f0 = 2, f1 = 2`
+
+    assert(math::linear_recurrence_matrix::get_nth_term_of_recurrence_series(
+               pell_recurrence, pell_base_case, 15) == 551614LL);
+    assert(math::linear_recurrence_matrix::get_nth_term_of_recurrence_series(
+               pell_recurrence, pell_base_case, 23) == 636562078LL);
+
+    /**
```
Comment (Panquesito7) [inline math/linear_recurrence_matrix.cpp:260]:
```suggestion
    /*
```

File: math/linear_recurrence_matrix.cpp [hunk 13]
```diff
@@ -0,0 +1,371 @@
...
+                                                    {0, 1, 1}},
+                                      trib_base_case = {
+                                          {0, 0, 1}};  // f0 = 0, f1 = 0, f2 = 1
+
+    assert(math::linear_recurrence_matrix::get_nth_term_of_recurrence_series(
+               tribonacci, trib_base_case, 11) == 149LL);
+    assert(math::linear_recurrence_matrix::get_nth_term_of_recurrence_series(
+               tribonacci, trib_base_case, 36) == 615693474LL);
+
+    /**
```
Comment (Panquesito7) [inline math/linear_recurrence_matrix.cpp:239]:
```suggestion
    /*
```

File: math/linear_recurrence_matrix.cpp [hunk 14]
```diff
@@ -0,0 +1,371 @@
...
+     * we can multiply B with itself n-1 times to obtain the required value
+     */
+    std::vector<std::vector<int64_t>> fibonacci_matrix = {{0, 1}, {1, 1}},
+                                      fib_base_case = {{0, 1}};
+
+    assert(math::linear_recurrence_matrix::get_nth_term_of_recurrence_series(
+               fibonacci_matrix, fib_base_case, 11) == 89LL);
+    assert(math::linear_recurrence_matrix::get_nth_term_of_recurrence_series(
+               fibonacci_matrix, fib_base_case, 39) == 63245986LL);
+    /**
```
Comment (Panquesito7) [inline math/linear_recurrence_matrix.cpp:212]:
```suggestion
    /*
```

File: math/linear_recurrence_matrix.cpp [hunk 15]
```diff
@@ -0,0 +1,371 @@
...
+}
+}  // namespace linear_recurrence_matrix
+}  // namespace math
+
+/**
+ * @brief Self test-implementations
+ * @returns void
+ */
+static void test() {
+    /**
```
Comment (Panquesito7) [inline math/linear_recurrence_matrix.cpp:192]:
```suggestion
    /*
```

File: math/linear_recurrence_matrix.cpp [hunk 16]
```diff
@@ -0,0 +1,371 @@
...
+    std::vector<std::vector<int64_t>>
+        custom_recurrence = {{1, 0, 1}, {0, 0, 1}, {0, 1, 2}},
+        custom_base_case = {{7, 2, 2}};  // `c = 7, f0 = 2, f1 = 2`
+
+    assert(math::linear_recurrence_matrix::get_nth_term_of_recurrence_series(
+               custom_recurrence, custom_base_case, 10, 1) == 18493LL);
+    assert(math::linear_recurrence_matrix::get_nth_term_of_recurrence_series(
+               custom_recurrence, custom_base_case, 19, 1) == 51531251LL);
+
+    /**
```
Comment (Panquesito7) [inline math/linear_recurrence_matrix.cpp:290]:
```suggestion
    /*
```

File: math/linear_recurrence_matrix.cpp [hunk 17]
```diff
@@ -0,0 +1,371 @@
...
+                                                             {1, 1, 1},
+                                                             {0, 0, 1}},
+                                      sum_fibo_base_case = {
+                                          {0, 1, 1}};  // `f0 = 0, f1 = 1`
+
+    assert(math::linear_recurrence_matrix::get_nth_term_of_recurrence_series(
+               sum_fibo_recurrence, sum_fibo_base_case, 13, 1) == 609LL);
+    assert(math::linear_recurrence_matrix::get_nth_term_of_recurrence_series(
+               sum_fibo_recurrence, sum_fibo_base_case, 16, 1) == 2583LL);
+    /**
```
Comment (Panquesito7) [inline math/linear_recurrence_matrix.cpp:324]:
```suggestion
    /*
```

**Review by Panquesito7 — APPROVED** [review]
Amazing work; I like how you've made & improved the code so well! 🎉
Code and documentation are pretty good and refined; LGTM. 😄 👍

**Review by ayaankhan98 — APPROVED** [review]
Great Work @AshishYUO @Panquesito7

**Pull request merged**